### PR TITLE
Fix errors revealed by Zygote's tests

### PR DESCRIPTION
**Pull request by oxinabox**
Found these because of Zygote's tests
while working on https://github.com/FluxML/Zygote.jl/pull/366

It is hard to add tests for them because they are about testing that
we didn't define a rule we should not have,
and that we don't thunk to many times

### CODE REVIEW

File: src/rulesets/Base/base.jl [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 @scalar_rule(abs(x::Real), sign(x))
 @scalar_rule(abs2(x), 2x)
-@scalar_rule(exp(x), Ω)
+@scalar_rule(exp(x::Real), Ω)
```
Comment (simeonschaub) [inline src/rulesets/Base/base.jl:7]: Is `exp` really the only complex function that causes problems with Zygote? Or are the others just not tested?

**Comment by willtebbutt** [comment]
Seems unlikely. Note [this issue](https://github.com/JuliaDiff/FiniteDifferences.jl/issues/75) on FiniteDifferences. [This](https://github.com/JuliaDiff/FiniteDifferences.jl/pull/76) PR aims to fix it.

File: src/rulesets/Base/base.jl (hunk 1 above)
Comment (oxinabox) [reply src/rulesets/Base/base.jl:7]: Testing is ongoing

Comment (simeonschaub) [reply src/rulesets/Base/base.jl:7]: I'm pretty sure the tests will fail for all our complex functions. Zygote does something quite different there.

**Comment by oxinabox** [comment]
Ok this is ready for review though it needs the fix in TestUtils before tests will run

File: src/rulesets/Base/base.jl (hunk 1 above)
Comment (oxinabox) [reply src/rulesets/Base/base.jl:7]: turns out it only need conjugating on input and output, and othereise all is well.

**Comment by oxinabox** [comment]
Once we have https://github.com/JuliaDiff/ChainRulesTestUtils.jl/pull/25
this is good to go.
(I tested and it passes on that branch)

File: src/rulesets/LinearAlgebra/dense.jl [hunk 2]
```diff
@@ -121,14 +121,11 @@ function rrule(::typeof(/), A::AbstractVecOrMat{<:Real}, B::AbstractVecOrMat{<:R
...
         # Optimization note: dAᵀ, dBᵀ, dC are calculated no matter which partial you want
-        # this is not a problem if you want the 2nd or 3rd, but if you want the first, it
-        # is fairly wasteful
-        _, dC = dC_pb(Ȳ)
-        _, dBᵀ, dAᵀ = dS_pb(extern(dC))
-
-        # need to extern as  dAᵀ, dBᵀ  are generally `Thunk`s, which don't support adjoint
-        ∂A = @thunk last(dA_pb(extern(dAᵀ)))
-        ∂B = @thunk last(dA_pb(extern(dBᵀ)))
+                _, dC = dC_pb(Ȳ)
```
Comment (nickrobinson251) [inline src/rulesets/LinearAlgebra/dense.jl:124]: Typo/tab here?

File: src/rulesets/Base/base.jl [hunk 3]
```diff
@@ -132,3 +132,13 @@ function rrule(::typeof(identity), x)
     end
     return x, identity_pullback
 end
+
+function rrule(::typeof(identity), x::Tuple)
+    # `identity(::Tuple)` returns multiple outputs;because that is how we think of
+    # returning a tuple, so its pullback needs to accept multiple inputs.
+    # `identity(::Tuple)` has one input, so its pullback should return 1 matching output
```
Comment (nickrobinson251) [inline src/rulesets/Base/base.jl:139]: Link to issue CRC on this?

Comment (oxinabox) [reply src/rulesets/Base/base.jl:139]:
```suggestion
    # `identity(::Tuple)` has one input, so its pullback should return 1 matching output
    # see https://github.com/JuliaDiff/ChainRulesCore.jl/issues/152
```

File: src/rulesets/LinearAlgebra/dense.jl (hunk 2 above)
Comment (oxinabox) [reply src/rulesets/LinearAlgebra/dense.jl:124]:
```suggestion
        _, dC = dC_pb(Ȳ)
```

File: src/rulesets/LinearAlgebra/dense.jl [hunk 4]
```diff
@@ -121,14 +121,11 @@ function rrule(::typeof(/), A::AbstractVecOrMat{<:Real}, B::AbstractVecOrMat{<:R
     C, dC_pb = rrule(adjoint, Cᵀ)
     function slash_pullback(Ȳ)
         # Optimization note: dAᵀ, dBᵀ, dC are calculated no matter which partial you want
-        # this is not a problem if you want the 2nd or 3rd, but if you want the first, it
-        # is fairly wasteful
         _, dC = dC_pb(Ȳ)
-        _, dBᵀ, dAᵀ = dS_pb(extern(dC))
+        _, dBᵀ, dAᵀ = dS_pb(unthunk(dC))
 
-        # need to extern as  dAᵀ, dBᵀ  are generally `Thunk`s, which don't support adjoint
```
Comment (nickrobinson251) [inline src/rulesets/LinearAlgebra/dense.jl:129]: why did this comment go?

Comment (oxinabox) [reply src/rulesets/LinearAlgebra/dense.jl:129]: It was wrong, we don't exten them

Comment (oxinabox) [reply src/rulesets/LinearAlgebra/dense.jl:129]: we `unthunk` them, but this is now a clear and standard pattern so pointing it out is redundant
things should in general be unthunked before being fed to pullbacks

Comment (nickrobinson251) [reply src/rulesets/LinearAlgebra/dense.jl:129]: good point! https://github.com/JuliaDiff/ChainRules.jl/issues/177

File: src/rulesets/Base/base.jl [hunk 5]
```diff
@@ -117,7 +118,7 @@ end
 
 function rrule(::typeof(*), x::Number, y::Number)
     function times_pullback(ΔΩ)
-        return (NO_FIELDS,  @thunk(ΔΩ * y'), @thunk(x' * ΔΩ))
+        return (NO_FIELDS,  @thunk(ΔΩ * y), @thunk(x * ΔΩ))
```
Comment (nickrobinson251) [inline src/rulesets/Base/base.jl:121]: this is functionally the same right? The `'` before was just a no-op

File: src/rulesets/LinearAlgebra/dense.jl [hunk 6]
```diff
@@ -40,17 +40,17 @@ end
 ##### `det`
 #####
 
-function frule((_, ẋ), ::typeof(det), x)
+function frule((_, ẋ), ::typeof(det), x::Union{Number, AbstractMatrix})
```
Comment (nickrobinson251) [inline src/rulesets/LinearAlgebra/dense.jl:43]: why the type annotations? (these do i believe cover all `det`/`logdet` methods in Base, but now not those extended by other packages, i suppose?) 

dose it warrant a comment in the code (or docs?) ?

File: src/rulesets/LinearAlgebra/structured.jl [hunk 7]
```diff
@@ -6,8 +6,15 @@
...
-    function Diagonal_pullback(ȳ)
-        return (NO_FIELDS, @thunk(diag(ȳ)))
+    function Diagonal_pullback(ȳ::AbstractMatrix)
+        return (NO_FIELDS, diag(ȳ))
+    end
+    function Diagonal_pullback(ȳ::Composite)
+        # TODO: Assert about the primal type in the Composite, It should be Diagonal
+        # infact it should be exactly the type of `Diagonal(d)`
+        # but right now Zygote loses primal type information so we can't use it.
+        # See https://github.com/FluxML/Zygote.jl/issues/603
```
Comment (nickrobinson251) [inline src/rulesets/LinearAlgebra/structured.jl:16]: i am confused by this: how do we end up with a `Composite` but not a `Composite{P}` for the correct `P`?

File: test/rulesets/LinearAlgebra/structured.jl [hunk 8]
```diff
@@ -6,7 +6,15 @@
         rrule_test(Diagonal, D, (randn(rng, N), randn(rng, N)))
         # Concrete type instead of UnionAll
         rrule_test(typeof(D), D, (randn(rng, N), randn(rng, N)))
+
+        # TODO: replace this with a `rrule_test` once we have that working
+        # see https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/24
+        res, pb = rrule(Diagonal, [1, 4])
+        @test pb(10*res) == (NO_FIELDS, [10, 40])
+        comp = Composite{typeof(res)}(;diag=10*res.diag)  # this is the structure of Diagonal
```
Comment (nickrobinson251) [inline test/rulesets/LinearAlgebra/structured.jl:14]:
```suggestion
        comp = Composite{typeof(res)}(; diag=10*res.diag)  # this is the structure of Diagonal
```

File: src/rulesets/Base/base.jl (hunk 5 above)
Comment (oxinabox) [reply src/rulesets/Base/base.jl:121]: On real numbers it was a noop,
on complex numbers it was AFAICT just wrong?

File: src/rulesets/LinearAlgebra/dense.jl (hunk 6 above)
Comment (oxinabox) [reply src/rulesets/LinearAlgebra/dense.jl:43]: Correct, I do not thing it deserves an annoation though,
it should be our practice to always include these when the primal method has them

File: src/rulesets/LinearAlgebra/structured.jl (hunk 7 above)
Comment (oxinabox) [reply src/rulesets/LinearAlgebra/structured.jl:16]: Because in the Zygote PR we turn all `NamedTuples` into `Composite{Any}`

**Comment by willtebbutt** [comment]
Cross referencing [this](https://github.com/FluxML/Zygote.jl/pull/366#issuecomment-620743107) -- I'm not sure why `abs2` doesn't work for complex numbers, but I can see why our current implementation of `abs` doesn't work.

File: src/rulesets/LinearAlgebra/dense.jl (hunk 6 above)
Comment (nickrobinson251) [reply src/rulesets/LinearAlgebra/dense.jl:43]: https://github.com/JuliaDiff/ChainRulesCore.jl/issues/155

File: src/rulesets/Base/base.jl (hunk 5 above)
Comment (nickrobinson251) [reply src/rulesets/Base/base.jl:123]: oh that makes much more sense (i never remember to think about Complex...)

Comment (oxinabox) [reply src/rulesets/Base/base.jl:123]: I wish i could forget about complex

Comment (oxinabox) [reply src/rulesets/Base/base.jl:123]: I have added test which fail without this change

**Pull request merged**